### ISSUE TITLE
MicroProfile Reactive Messaging 3.0 Support

### ISSUE THREAD

**Issue by Emily-Jiang** [OpenLiberty/open-liberty#19889]
## Description

The current MicroProfile Reactive Messaging 1.0 does not work with MicroProfile 4.x. Customers that use RM 1.0 are stuck. Quite a few customers asked about when RM 2.0 will be supported in Open Liberty many times (I was asked at JChampion conference 2022 on Friday 21st Jan 2022). 

We should directly support the recent release of RM 3.0 (Jakarta EE 9.1 alignment) to make it work with Jakarta EE 10 and MP 6.0.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Documents

When available, add links to required feature documents. Use "N/A" to mark particular documents which are not required by the feature.

- Aha: Externally raised RFE ([Aha](https://cloud-platform.ideas.aha.io/))
  - Link the RFE with this issue
- UFO: https://ibm.box.com/s/bffbods91oeddvxgbb5gpodq6m8x4ymy
- FTS: (https://github.com/OpenLiberty/open-liberty/issues/25853)
- Beta Blog: (https://github.com/OpenLiberty/open-liberty/issues/26642)
- GA Blog: (https://github.com/OpenLiberty/open-liberty/issues/27213)

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Process Overview

- [Prioritization](#prioritization)
- [Design](#design)
- [Implementation](#implementation)
- [Legal and Translation](#legal-and-translation)
- [Beta](#beta)
- [GA](#ga)
  - [Focal Point Approvals](#focal-point-approvals-complete-by-feature-complete-date)
- [Other Deliverables](#other-deliverables)

### General Instructions

The process steps occur roughly in the order as presented. Process steps occasionally overlap.

Each process step has a number of tasks which must be completed or must be marked as not applicable ("N/A").

Unless otherwise indicated, the tasks are the responsibility of the Feature Owner or a Delegate of the Feature Owner.

If you need assistance, reach out to the [OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect).

**Important: Labels are used to trigger particular steps and must be added as indicated.** 

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Prioritization** (Complete Before Development Starts)

The ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) and area leads are responsible for prioritizing the features and determining which features are being actively worked on.

### **Prioritization**
- [x] Feature added to the "New" column of the [Open Liberty project board](https://github.com/orgs/OpenLiberty/projects/2)
  - Epics can be added to the board in one of two ways:
    - From this issue, use the "Projects" section to select the appropriate project board.
    - From the appropriate project board click "Add card" and select your Feature Epic issue
- [x] Priority assigned
  - Attend the Liberty Backlog Prioritization meeting

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Design** (Complete Before Development Starts)

Design preliminaries determine whether a formal design, which will be provided by an [Upcoming Feature Overview (UFO)](https://ibm.box.com/v/UFO-Template) document, must be created and reviewed.  A formal design is required if the feature requires any of the following: UI, Serviceability, SVT, Performance testing, or non-trivial documentation/ID.

### **Design Preliminaries**
- [ ] UI requirements identified. (Owner and [UI focal point](https://github.com/orgs/OpenLiberty/teams/ui-approvers))
- [x] ID requirements identified. (Owner and [ID focal point](https://github.com/orgs/OpenLiberty/teams/id-approvers))
   - Refer to [Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty).
   - Feature Owner adds label `ID Required`, if non-trivial documentation needs to be created by the ID team.
   - ID adds label `ID Required - Trivial`, if no design will be performed and only trivial ID updates are needed.
- [x] Serviceability Requirements Identified. (Owner and [Serviceability focal point](https://github.com/orgs/OpenLiberty/teams/serviceability-approvers))
- [x] SVT Requirements Identified. (Owner and [SVT focal point](https://github.com/orgs/OpenLiberty/teams/svt-approvers))
- [x] Performance testing requirements identified. (Owner and [Performance focal point](https://github.com/orgs/OpenLiberty/teams/performance-approvers))

### **Design**
- [x] POC Design / UFO review requested.
  - Owner adds label `Design Review Request`
- [x] POC Design / UFO review scheduled.
  - Follow the instructions in POC-Forum repo
- [x] POC Design / UFO review completed.
- [x] POC / UFO Review follow-ons completed.
- [x] POC Design / UFO approval requested.
  - Owner adds label `Design Approval Request`
- [x] Design / UFO approved. ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) or N/A
  - ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) adds label `Design Approved`
  - Add the public link to the UFO in Box to the [Documents](#documents) section.
  - The UFO must always accurately reflect the final implementation of the feature. Any changes must be first approved. Afterwards, update the UFO by creating a copy of the original approved slide(s) at the end of the deck and prepend "OLD" to the title(s). A single updated copy of the slide(s) should take the original's place, and have its title(s) prepended with "UPDATED".

### **No Design**
- [ ] No Design requested.
  - Owner adds label `No Design Approval Request`
- [ ] No Design / No UFO approved. ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) or N/A
  - Approver adds label `No Design Approved`
- [ ] Feature / Capability stabilization or discontinuation or N/A
  - Owner adds label `Product Management Approval Request` and notifies [OpenLiberty/product-management](https://github.com/orgs/OpenLiberty/teams/product-management)
  - Approver adds label `Product Management Approved` ([OpenLiberty/product-management](https://github.com/orgs/OpenLiberty/teams/product-management))
  - Note: For stabilized, superseded, and discontinued feature/capability, skip the [Beta](#beta) section of the template (you may delete it).  Otherwise, proceed as normal.

### **FAT Documentation**
- [x] "Feature Test Summary" child task created
  - Use the [Feature Test Summary Template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=)
  - Add FTS issue link to the [Documents](#documents) section.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Implementation**

A feature must be [prioritized](https://github.com/orgs/OpenLiberty/projects/2) before any implementation work may begin to be delivered (inaccessible/no-ship).  However, a design focused approach should still be applied to features, and developers should think about the feature design prior to writing and delivering any code.  
Besides being prioritized, a feature must also be socialized (or No Design Approved) before any beta code may be delivered.  All new Liberty content must be inaccessible in our GA releases until it is [Feature Complete](#feature-complete) by either marking it `kind=noship` or [beta fencing](#beta-code) it.  
Code may not GA until this feature has obtained the "Design Approved" or "No Design Approved" label, along with all other tasks outlined in the [GA](#ga) section.

### **Feature Development Begins**
- [x] Add the `In Progress` label
- [x] #25385
- [x] #25386
- [x] #25388
- [x] #25389
- [x] #25390
- [x] #25423
- [x] #25424
- [x] #25450
- [x] #25644
- [x] #25645
- [x] #25852
- [x] #25969
- [x] #25970
- [x] #25975
- [x] #26082
- [x] #26204
- [x] #26220
- [x] #26243
- [x] #26274
- [x] #26352
- [x] #26353
- [x] #26428
- [x] #26664 
- [x] #26665
- [x] #26678
- [x] #27189 
- [x] #26692
- [x] #26852

## **Legal and Translation**

In order to avoid last minute blockers and significant disruptions to the feature, the **legal items need to be done as early in the feature process as possible**, either in design or as early into the development as possible.  Similarly, translation is to be done concurrently with development.  Both **MUST** be completed before Beta or GA is requested.

### **Legal** (Complete before Feature Complete Date)
- [x] Changed or new open source libraries are cleared and approved, or N/A. (Legal Release Services/Cass Tucker/Release PM).
- [x] Licenses and Certificates of Originality (COOs) are updated, or N/A

### **Translation** (Complete 1 week before Feature Complete Date)
- [x] PII updates are merged, or N/A. Note timing with translation shipments.

### **Innovation** (Complete 1 week before Feature Complete Date)
- [x] Consider whether any aspects of the feature may be patentable. If any identified, disclosures have been submitted.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Beta**

In order to facilitate early feedback from users, all new features and functionality should first be released as part of a beta release.

### **Beta Code**
- [x] Beta fence the functionality
  - `kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`
- [x] Beta development complete and feature ready for inclusion in a beta release
  - Add label `target:beta` and the appropriate `target:YY00X-beta` (where YY00X is the targeted beta version).
- [x] Feature delivered into beta
  - ([OpenLiberty/release-manager](https://github.com/orgs/OpenLiberty/teams/release-manager)) adds label `release:YY00X-beta` (where YY00X is the first beta version that included the functionality).

### **Beta Blog** (Complete 1.5 weeks before beta eGA)
- [x] Beta blog issue created and populated using the [Open Liberty BETA blog post](https://github.com/OpenLiberty/open-liberty/issues/new/choose) template.
  - Add a link to the beta blog issue in the [Documents](#documents) section.
  - Note: This is for inclusion into the overall [beta release blog post](https://openliberty.io/blog/?search=beta&key=tag).  If, in addition, you'd also like to create a dedicated blog post about your feature, then follow the "Standalone Feature Blog Post" instructions under the [Other Deliverables](#other-deliverables) section.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **GA**

A feature is ready to GA after it is Feature Complete and has obtained all necessary Focal Point Approvals. 

### **Feature Complete**
- [ ] Feature implementation and tests completed.
  - [ ] All PRs are merged.
  - [x] All related/child issues are closed.
  - [x] All stop ship issues are completed.
- [x] Legal: all necessary approvals granted.
- [x] Translation: Feature may only proceed to GA if it has either `Translation - Complete` or `Translation - Missing` label
  - If all translation has been delivered to `release` branch, feature owner adds label `Translation - Complete`.
  - If missing translation does not cause a break in functionality, nor a security or production outage risk, feature owner adds label `Translation - Missing`.
    - Once all missing translations are delivered, the `Translation - Missing` label is replaced with `Translation - Complete`.
  - If missing translation could cause a break in functionality or a security or production outage risk, feature owner adds the `Translation - Blocked` label.
    - Featues with `Translation - Blocked` may **NOT** proceed to GA until the label has been replaced with either `Translation - Missing` or `Translation - Complete`.
  - For further guidance, contact [Globalization focal point](https://github.com/orgs/OpenLiberty/teams/globalization-approvers) or the [Release Architect](https://github.com/orgs/OpenLiberty/teams/release-architect).
- [x] GA development complete and feature ready for inclusion in a GA release
  - Add label `target:ga` and the appropriate `target:YY00X` (where YY00X is the targeted GA version).
  - Inclusion in a release requires the completion of all Focal Point Approvals.

### **Focal Point Approvals** (Complete by Feature Complete Date)

These occur only after GA of this feature is requested (by adding a `target:ga` label).  GA of this feature may not occur until all approvals are obtained.

### **All Features**
- [x] **APIs/Externals** Externals have been reviewed or N/A.  ([OpenLiberty/externals-approvers](https://github.com/orgs/OpenLiberty/teams/externals-approvers))
  - Approver adds label `focalApproved:externals` 
- [x] **Demo** Demo is scheduled for an upcoming EOI or N/A. ([OpenLiberty/demo-approvers](https://github.com/orgs/OpenLiberty/teams/demo-approvers))
  - Add comment `@OpenLiberty/demo-approvers Demo scheduled for EOI [Iteration Number]` to this issue.
  - Approver adds label `focalApproved:demo`.
- [x] **FAT** All Tests complete and running successfully in SOE or N/A. ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers))
  - Approver adds label `focalApproved:fat`.

### **Design Approved Features**
- [x] **ID** - Documentation is complete or N/A. ([OpenLiberty/id-approvers](https://github.com/orgs/OpenLiberty/teams/id-approvers))
  - Approver adds label `focalApproved:id`.
  - > **_NOTE:_**  If only trivial documentation changes are required, you may reach out to the ID Feature Focal to request a `ID Required - Trivial` label.  Unlike features with regular ID requirement, those with `ID Required - Trivial` label do not have a hard requirement for a Design/UFO.
- [x] **InstantOn** - InstantOn capable or N/A. ([OpenLiberty/instantOn-approvers](https://github.com/orgs/OpenLiberty/teams/instantOn-approvers))
  - Approver adds label `focalApproved:instantOn`.
- [x] **Performance** - Performance testing is complete or N/A. ([OpenLiberty/performance-approvers](https://github.com/orgs/OpenLiberty/teams/performance-approvers))
  - Approver adds label `focalApproved:performance`.
- [x] **Serviceability** - Serviceability has been addressed or N/A. ([OpenLiberty/serviceability-approvers](https://github.com/orgs/OpenLiberty/teams/serviceability-approvers))
  - Approver adds label `focalApproved:sve`.
- [x] **STE** - Skills Transfer Education chart deck is complete or N/A. ([OpenLiberty/ste-approvers](https://github.com/orgs/OpenLiberty/teams/ste-approvers))
  - Approver adds label `focalApproved:ste`.
- [x] **SVT** - System Verification Test is complete or N/A. ([OpenLiberty/svt-approvers](https://github.com/orgs/OpenLiberty/teams/svt-approvers))
  - Approver adds label `focalApproved:svt`.

### **Remove Beta Fencing** (Complete by Feature Complete Date)
- [x] Beta guards are removed, or N/A
  - Only after all necessary Focal Point Approvals have been granted.

### **GA Blog** (Complete by Feature Complete Date)
- [x] GA Blog issue created and populated using the [Open Liberty GA release blog post](https://github.com/OpenLiberty/open-liberty/issues/new/choose) template.
  - Add a link to the GA Blog issue in the [Documents](#documents) section.
  - Note: This is for inclusion into the overall [release blog post](https://openliberty.io/blog/?search=release&key=tag).  If, in addition, you'd also like to create a dedicated blog post about your feature, then follow the "Standalone Feature Blog Post" instructions under the [Other Deliverables](#other-deliverables) section.

### **Post GA**
- [ ] Replace `target:YY00X` label with the appropriate `release:YY00X`. ([OpenLiberty/release-manager](https://github.com/orgs/OpenLiberty/teams/release-manager))

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Other Deliverables**

- [ ] **Standalone Feature Blog Post** A blog post specifically about your feature or N/A.  ([OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect))
  - This should be strongly considered for larger or more prominent features.
  - Follow [instructions](https://github.com/OpenLiberty/blogs/tree/draft#writing-and-publishing-blog-posts-on-the-openlibertyio-blog) in the blogs repo.
- [ ] **OL Guides** OL Guides assessment is complete or N/A. ([OpenLiberty/guide-assessment](https://github.com/orgs/OpenLiberty/teams/guide-assessment/members))
- [ ] **Dev Experience** Developer Experience & Tools work is complete or N/A. ([OpenLiberty/dev-experience-assessment](https://github.com/orgs/OpenLiberty/teams/dev-experience-assessment/members))
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

**Comment by inad9300** [comment]
It's great to see this in the pipeline. Any estimation on when it will be available? What's the recommended solution in the meantime?

**Comment by Emily-Jiang** [comment]
We constantly revisit our features in the pipeline. For this feature, it is not high up in the pipeline as yet at the moment, due to resource constraints. Do you currently use MicroProfile reactive messaging 1.0 in the production? Unfortunately, the workaround is to stay with the older version of MicroProfile release if you want to use MicroProfile Reactive Message together with other MicroProfile features.

**Comment by inad9300** [comment]
Thanks for your prompt answer. We are not currently using this feature in production, but rather performing an evaluation of technologies for a new project. Unfortunately, I don't think it makes sense to begin a new project using MicroProfile 3.x, being two major versions behind now. Is there any other known workaround? For example, could we modify our `pom.xml` and depend directly on `microprofile-reactive-messaging-api` and `smallrye-reactive-messaging-provider`, or do you foresee any big issues with such an approach?

**Comment by jhanders34** [comment]
My notes from the UFO socialization that was done on September 8:

**Java APIs / SPIs** (page 17)
- [x] Change metrics to mpMetrics in the second bullet

**Emitter and @Channel** (page 19)
- [x] Update the right example to be lower so you can see the word `Message`

**Emitter and Channel Usage** (page 21)
- [x] Example should be updated.  `toUpperCase` returns a value, but there is no `@Outgoing` so it can't return String or it needs `@Outgoing` added.

**Backpressure Strategy** (page 24)
- [x] Make sure that the 128 message buffer default is highlighted better.

**Negative Acknowledgment (Nack)** (page 28)
- [x] POST_PROCESSING implementation does the ack / nack.  Make that more clear that implementation = Liberty.  Maybe just added `(Liberty)` after the word `implementation`.

**Admin / Config / Command Line** (page 40)
- [x] Provide details about what the sensible defaults are
- [x] Confirm validity of liberty kafka connection bootstrap server line.  `bootstrap.servers` vs `bootstrap.server`.  Can provide multiple hosts/ports.  I believe on the call we confirmed that it is `servers` and not `server` in the existing guides.

**Automated Testing** (page 46)
- [x] Clarify how both user provided kafka connectors and the built-in liberty kafka connectors are covered.  FATs I believe cover one and the TCK the other.
- [x] FAT test for messages coming out for when validation fails for missing mp config properties for different configurations

**Platform / cloud considerations** (page 49)
- [x] For Instant On, should have a statement that we will need to not start things that depend on the MP Config properties until restore.  Open a GitHub issue for the Instant On folks to prioritize mpReactiveMessaging features enablement for Instant On so they can discuss on how to do this in the future outside of this UFO.  

**Serviceability** (page 52)
- [x] Update about messages for validation errors.
- [x] Include a message when things get "stuck" when things are not acknowledged when MANUAL acknowledgement is configured and it is not done.

**Comment by NottyCode** [comment]
Slide 23 indicates a strategy of BUFFER has a limit, but the Javadoc from the screenshot indicates that all messages are buffered, which makes UNBOUNDED_BUFFER and BUFFER seem identical. This is perhaps a Javadoc bug from upstream?

How does this feature handle changes to the configuration?

In what way does the configuration processing of this feature block the possibility of InstantOn working?

**Comment by abutch3r** [comment]
@NottyCode 

The upstream Javadoc short descriptions as seen on slide 23 are the issue here. When looking at the full definitions in the javadoc it does cover the differences between the two.
![image](https://github.com/OpenLiberty/open-liberty/assets/21243172/d88e39cd-3732-4779-8b7a-4d10718b69d1)

The feature does not support changes to its configuration once the application has started for either Reactive Messaging or our Connector. All Configuration is read on App Start.

As the configuration is read only once on App start, changes, which we do highly expect there to be, especially details for connecting to kafka. Then currently the `afterApp` option currently does cause the app to fail to checkpoint. Deferring the processing of the configuration to after app restore is currently being investigated. The Smallrye implementation passes the connector config to the `liberty-kafka` connector, so if the implementation startup is deferred, then the connector will be as well.

**Comment by Azquelt** [comment]
There are a couple of problems that have arisen during development which are going to require changes in the design:

- The Kafka connector currently tries to batch up message acknowledgements to improve performance, and does not mark an acknowledgement as complete until the batch is sent. However, some parts of reactive messaging wait for acknowledgement of one message to complete before processing the next message.
  - This results in slow processing of messages in RM 1.0 and the whole message stream stalling in RM 3.0
  - In order to work well with the rest of reactive messaging, if the Kafka connector batches messages, it must tell RM that the acknowledgement is complete when it receives it, even if it hasn't sent the batch yet.
  - I propose adding a new connector option `<channel prefix>.fast.ack` to change this behaviour
    - it will default to `true` in RM 3.0 so that it works better with the rest of RM
    - it will default to `false` in RM 1.0 to avoid breaking existing users who may rely on the current behaviour and to maintain zero migration
  - Related issue: #26692
- During review of another issue, the concurrency team pointed out some issues which also apply to the Kafka container
  - The Kafka container processes messages asynchronously but needs to ensure that the correct application context and TCCL are set on the thread.
  - Currently it uses a `ManagedExecutorService` to do this but there are two problems:
    - we don't allow the user to configure _which_ `ManagedExecutorService` is used
    - this requires a dependency on the `concurrent-x.x` feature
  - I propose making the following changes:
    - If `concurrent-x.x` is not enabled, use the built-in `WSContextService` to transfer the appropriate thread context
    - If `concurrent-x.x` is enabled, use the default `WSContextService` (which the user can configure)
    - Add a config option to allow the user to set the context service to use
      - server.xml element `<mpReactiveMessaging contextServiceRef="xyz"/>` would set the context service for all Kafka connections
      - connector configuration option `<channel prefix>.context.service="xyz"` would set the context service for a specific channel and would override any configuration from the server.xml because it is more specific
  - Related issue: #26678

I will update the UFO with these changes. Since it's already been reviewed once, I will mark the changed slides as NEW and keep the original side at the end of the presentation marked as OLD.

**Comment by Azquelt** [comment]
@NottyCode I've updated the UFO with the changes in the comment above. **The new/updated slides are 8, 11, 12, 43-45 and 61** (going by the slide numbers in the corners which doesn't match the box preview numbering because the presentation has hidden slides)

**Comment by jwalcorn** [comment]
Any updated outlook on this one?  It had been targeted for 23.0.0.12, but I see that target got removed recently.  Has this slipped out to 2024?  I tend to only move to the LTS releases, so maybe 24.0.0.3?  And I assume this will require the big package change from `javax.*` to `jakarta.*`?

**Comment by tevans78** [comment]
@jwalcorn Yes, we are now targeting 24.0.0.1 and yes it will be `jakarta.*` based (compatible with MP 5.0, 6.0 and 6.1).

**Comment by Emily-Jiang** [comment]
2nd UFO review comments:
Slide 40
1. Remove the server.xml configuration for ContextService for the property `context.service`
2. Update the slides to mention the connector config and channel config. For connector config, specify the full name or clarify the naming convention for the connector config.
3. Specify the priority order if both configs are present. 
4. Specify that the config will be ignored for a channel that does not support async.

**Comment by Azquelt** [comment]
Point 4 above is not correct. A connector will receive all the config options that the user passes to it. It's up to individual connectors to decide how to handle properties they don't understand.

**Update**: We discussed this and will update the slides so that it's clear that the `context.service` option only applies to the `liberty-kafka` connector.

**Comment by abutch3r** [comment]
@OpenLiberty/ste-approvers STE education is available at https://ibm.ent.box.com/file/1412639185359 please can you review the document and provide any feed

**Comment by abutch3r** [comment]
- UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? Have these issues been addressed in the implementation?
  - The UFO provides where we know most of the issues arise - the kafka connector
    - These errors are exposed to the logs via the Liberty connector or via the underlying client 
  - The UFO covers changes made to the underlying implementation to better expose configuration errors and how these help diagnose issues
 
- Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).
    a) What problem paths were tested and demonstrated?
    - Kafka acknowledgement
    - Kafka connections
    - Configuration errors
    b) Who did you demo to?
      - Local team have been demo'd to
      - Wider Demo is occured for Tuesday 6th of February and will include a few key error paths
    c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?
      - Yes

- SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
    a) Who conducted SVT tests for this feature?
      - Brian Hanczaryk
    b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?
      - Yes, SVT agrees that the serviceability of any problem encountered was sufficient to avoid PMRs or L2 should be able to quickly address those problems without engaging L3.

- Which L2 / L3 queues will handle PMRs for this feature? Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it. Ask Don Bourne if you need links or more info.
  - WAS L3: CDI
- Does this feature add any new metrics or emit any new JSON events? If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?
  - New Metric added - listed as requiring an update as part of the RM/RSO docs - https://github.com/OpenLiberty/docs/issues/6928

**Comment by tngiang73** [comment]
@abutch3r : WASWIN L2 is good with STE slides. Thanks!

**Comment by abutch3r** [comment]
@OpenLiberty/demo-approvers Demo scheduled for EOI 24.03
